doc/man/dune-copasi.1
.TH man 1 "02 August 2023" "2.0" "dune-copasi man page"
.SH NAME
dune-copasi \- Diffusion-Reaction numerical simulator
.SH SYNOPSIS
dune-copasi [OPTIONS]
.SH DESCRIPTION
dune-copasi is a numerical simulator for diffusion-reaction systems on single or multiple compartments.
.SH BUGS
https://gitlab.dune-project.org/copasi/dune-copasi/-/issues
.SH AUTHOR
Santiago Ospina De Los Rios (santiago.ospina@iwr.uni-heidelberg.de)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
Guiding Principles

    Changelogs are for humans, not machines.
    There should be an entry for every single version.
    The same types of changes should be grouped.
    Versions and sections should be linkable.
    The latest version comes first.
    The release date of each version is displayed.
    Mention whether you follow Semantic Versioning.

Types of changes

    Added         for new features.
    Changed       for changes in existing functionality.
    Deprecated    for soon-to-be removed features.
    Removed       for now removed features.
    Fixed         for any bug fixes.
    Security      in case of vulnerabilities.
 -->

## [Unreleased] ([git-diff][Unreleased-diff])
### Added
 - Doxygen documentation is now deployed to Netlify !73
### Changed
 - Docusaurus is updated to version [`2.0.0-alpha.75`](https://github.com/facebook/docusaurus/blob/master/CHANGELOG.md#200-alpha75-2021-04-30) !72

## [1.1.0] ([git-diff][1.1.0-diff]) - 2021-04-29
### Added
 - Dune options file receives `CMAKE_OPTIONS` and `MAKE_OPTIONS` !60
 - Parameter to use max norm on newton steps !65
 - Parameter to accept best solution in linear search on newton steps !65
 - Event timestepper !64
 - Preprocessor macros containing version information !67
### Changed
 - TIFF helper is compiled in the dunecopasi library !62
 - Use `NewtonMethod` instead of `Newton` class from PDELab !65
### Fixed
 - Executables can be compiled without the library !60
 - All dune cmake flags are now transitively passed to dune-copasi targets !60
 - System tests on main executables were ignored !66
 - Snap to final time now avoids very small timesteps !61
 - CMake config file was installed on the wrong path !68

## [1.0.0] ([git-diff][1.0.0-diff]) - 2021-02-11
### Added
- Custom membrane flux !30
- Skip intersection methods in local operators !36
- Parameters to control Newton's method !30
- Stepper interface to snap solution to a specific time !45
- Provide help message when executables are not used correctly !49
- Produce Debian Packages on the CI !49
- Installation is now divided on three components: `Runtime|Library|Development` !49
- Define a recommended dune options file [`dune-copasi.opts`](dune-copasi.opts) !49
- Versioned documentation !49
- Dependency on `pkg-config` !52
### Changed
- TIFF images are clamped instead of zeroed when evaluated outside its domain !47
- Produce versioned documentation !49
- Travis and Appveyor are replaced for GitHub Actions on `Linux|MacOS|Windows` !49
- Increment CMake required version to 3.13 !49
- Installed CMake project is now consumable by other CMake projects !49
- Vendor GHC with CMake instead of git submodules !49
- Increase vendored GHC version to 10.5 !49 & !51
- Force GHC usage whenever c++ filesystem cannot be found, else, optional !49
- CI scripts are improved to simplify usage !49
- Simplify build and usage on docker containers !49
- Drastically reduce size of final docker container !49
- Improve performance on cases with no interaction between all species !43
- `dune-logging` and `dune-multidomaingrid` no longer require a COPASI namespace fork !53
### Removed
- Automatic flux between compartment components with same name !30
- Jacobian operator is managed by the stepper instead of the model !39
- Jacobian tests are improved and extended to cover more cases !39
- CMake module to find muparser, use `pkg-config` instead !52
### Fixed
- Finished and documented Installation procedure !49
- Wrong rounding on y pixels on TIFF file reads !47
- Improper vector allocation in multidomain intersections !39
- Final timestep reduction to reach `end_time` failed in adaptive stepper !43
- Infinite loop when final adaptive step failed !45
- Binary executables are now installed !49
- Support for shared libraries !54

## [0.3.0] ([git-diff][0.3.0-diff]) - 2020-10-07
### Added
- User documentation !28
- Scape VTK write by omitting `[writer]` section !24
- Added Explicit and Diagonally Runge-Kutta time solvers !25
- 3D simulations are now possible !27
### Changed
- Improved errors when a config file is incorrect !25
- Update meaningful logging values on the test ini file !25
- Clean up logger output !25
- Time stepping is now done outside the model class !25
- Writer might write on different files on request !25
- Move read/write responsibility to model state !25
- States are now independent from models !25
- Output files changed its name scheme !25
- Configuration file changed its structure (see git diff on tests) !25
### Removed
- Coefficient mapper and its use on local operators !25
- Remove the operator map of objects within the model class !25
### Fixed
- Violation of the One-Definition-Rule due to external linkage of `LocalOperatorApply` lambdas !23
- Error on the gaussian equation used on comparisons !23

## [0.2.0] ([git-diff][0.2.0-diff]) - 2020-02-20
### Added
- Code documentation
- Data Context concept for factories
- Factory concept for arbitrary object instantiation
- Add factories for finite element and finite element map
- Brief installation instructions
- Models can interpolate grid functions
- Grid utilities to recognize and mark tripes of entities
- A finite volume loca operator
- Grid function getters for external use
- Single domain executable
### Changed
- Move and rename header files
- Multidomain finite element was split into a multidomain and a dynamic power finite element
- Code DUNE dependencies are set to release 2.7 (See README.md)
- Other DUNE dependencies are set to COPASI/support/dune-copasi (See README.md)
- Executable is an optional build
- Library is optional and is split into smaller libraries
- Bump version utility updated to python3
### Fixed

src/dune-copasi.cc
#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <dune/copasi/gmsh_reader.hh>
#include <dune/copasi/model_diffusion_reaction.hh>
#include <dune/copasi/model_diffusion_reaction.cc>
#include <dune/copasi/model_multidomain_diffusion_reaction.hh>
#include <dune/copasi/model_multidomain_diffusion_reaction.cc>

#include <dune/grid/multidomaingrid.hh>
#include <dune/grid/io/file/gmshreader.hh>

#include <dune/logging/logging.hh>

#include <dune/common/exceptions.hh>
#include <dune/common/parametertree.hh>
#include <dune/common/parametertreeparser.hh>
#include <dune/common/parallel/mpihelper.hh>

#include <iostream>

int main(int argc, char** argv)
{

  try{
    // initialize mpi
    auto& mpi_helper = Dune::MPIHelper::instance(argc, argv);
    auto comm = mpi_helper.getCollectiveCommunication();

    // Read and parse ini file
    if (argc!=2)
      DUNE_THROW(Dune::IOError, "Wrong number of arguments");
    const std::string config_filename = argv[1];

    Dune::ParameterTree config;
    Dune::ParameterTreeParser ptreeparser;
    ptreeparser.readINITree(config_filename, config);

    // initialize loggers
    Dune::Logging::Logging::init(comm,config.sub("logging"));

    using namespace Dune::Literals;
    auto log = Dune::Logging::Logging::logger(config);
    log.notice("Starting dune-copasi"_fmt);

    // create a grid
    constexpr int dim = 2;
    using HostGrid = Dune::UGGrid<dim>;
    using MDGTraits = Dune::mdgrid::DynamicSubDomainCountTraits<dim,1>;
    using Grid = Dune::mdgrid::MultiDomainGrid<HostGrid,MDGTraits>;

    auto& grid_config = config.sub("grid");
    auto level = grid_config.get<int>("initial_level",0);

    log.info("Creating a rectangular grid in {}D"_fmt, dim);

    auto grid_file = grid_config.get<std::string>("file");

    auto [grid_ptr,host_grid_ptr] = Dune::Copasi::GmshReader<Grid>::read(grid_file,config);

    log.debug("Applying global refinement of level: {}"_fmt, level);
    grid_ptr->globalRefine(level);

    auto& model_config = config.sub("model");
    Dune::Copasi::ModelMultiDomainDiffusionReaction<Grid> model(grid_ptr,model_config);
    model.run();

    return 0;
  }
  catch (Dune::Exception &e){
    std::cerr << "Dune reported error: " << e << std::endl;
  }
  catch (...){
    std::cerr << "Unknown exception thrown!" << std::endl;
  }
}
